proposals/0000-tuple-syntax.rst

```diff
@@ -0,0 +1,414 @@
+#. An instance declaration like ``instance (C a, C b) => C (Tuple [a, b]) where ...`` would be
+   rejected because it uses a type family in the instance head. We might choose to relax
+   this restriction, by allowing type families in an instance head, as long as they can
+   reduce to a ground (i.e. type-family-free) type. This proposal does *not* include such
+   a lifting of the restriction, as the workaround is straightforward: just write
+   ``instance (C a, C b) => C (Tuple2 a b) where ...``. Still, we may decide to revisit
+   this in the future.
+
+#. In due course, we may wish to consider re-exporting some of the definitions
+   above from modules not in the ``GHC.`` namespace, perhaps even including the
+   ``Prelude``. This proposal does *not* make any such suggestions, and it does *not*
+   depend on any such ideas being adopted in the future. Any such idea would
+   be evaluated by the Core Libraries Committee independently of this proposal.
+
+#. This proposal changes the name of the constructor of the unary boxed tuple ``Solo``,
+   from ``Solo`` to ``MkSolo``. The proposal includes a deprecated ``Solo`` pattern
+   synonym to enable a migration period.
+
+Costs and Drawbacks
+-------------------
+1. This is one more feature to maintain, but the code would be pretty local.
+
+#. Having multiple ways of naming one thing may offer a boon to *writers* of code
+   (they can choose whichever way to name a tuple that they like), but it imposes
+   a burden on *readers* of code, who may need to be familiar with all possible
+   ways of describing a tuple (and that they are interchangeable). Careful
+   documentation of these ideas -- ideally, in the Haddock documentation for the
+   names introduced above -- will help to mitigate this problem.
+
+#. A particular class of code readers are beginners, and having multiple different
+   ways to say the same thing is particularly challenging for beginners. We should
+   thus think carefully about how to present these names to beginners, if
+   ``-XNoListTuplePuns`` catches on.
+
+Alternatives
+------------
+1. Instead of defining ``TupleN`` as a type family (as done here), it could be
+   a data family, effectively replacing the ``Tuple2``, ``Tuple3``, ..., definitions.
+   This design would seem to be too complicated to be the primitive definition
+   of tuples, however, when a very vanilla datatype like ``data Tuple2 a b = (a, b)``
+   suffices.
+
+#. Instead of introducing new names, we could use more mixfix bits of punctuation,
+   such as ``(~ ty1, ty2 ~)`` for normal tuples and ``(% ty1, ty2 %)`` for constraint
+   tuples. This was not as popular in a recent `straw poll <https://github.com/ghc-proposals/ghc-proposals/pull/458#issuecomment-982230541>`_.
+
+#. Controlling the ``(# ... | ... | ... #)`` syntax for unboxed sum types with
+   ``-XNoListTuplePuns`` is not necessary to avoid punning, but is done only for
+   consistency. We could skip this, but I prefer keeping it as proposed.
```

Comment:
This is an interesting case. It's not necessary, so if the choice is given, would one prefer using `(# a | b | c #)` or `Sum# [a, b, c]`? Would removing this syntax be considered an anti-feature?

If programmers prefer the former, then it raises an interesting question. It is arguable that unboxed sum are niche enough a feature not to deserve taking up syntactic real estate. However, tuples definitely are. So it would argue that the alternative, above, proposing a special syntax for tuples may have more weight to it than it first seems.

Reply:
I'd be quite happy with special syntax for tuples. But this seems to be the minority position, if https://github.com/ghc-proposals/ghc-proposals/pull/458#issuecomment-982230541 is to be believed.